sdk/java/pulumi/src/main/java/com/pulumi/core/Urn.java

```diff
@@ -1,75 +1,312 @@
 package com.pulumi.core;
 
-import com.google.common.base.Strings;
-import com.pulumi.deployment.Deployment;
-import com.pulumi.resources.Resource;
+import com.pulumi.core.internal.Optionals;
+import com.pulumi.core.internal.annotations.InternalUse;
 
-import javax.annotation.Nullable;
 import javax.annotation.ParametersAreNonnullByDefault;
-import java.util.Arrays;
 import java.util.Objects;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+import static com.pulumi.core.internal.Objects.require;
+import static com.pulumi.core.internal.Strings.isNonEmptyOrNull;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
 
 /**
  * An automatically generated logical URN, used to stably identify resources. These are created
  * automatically by Pulumi to identify resources. They cannot be manually constructed.
+ * <p>
```

Comment:
Consider linking to the docs https://www.pulumi.com/docs/intro/concepts/resources/names/#urns that defines what it is.